### PR TITLE
Fix UBSan vptr UB: wrap IDxcBlob as ISlangBlob instead of reinterpret-casting

### PR DESCRIPTION
**Pull request by jvepsalainen-nv**
## Summary

Fixes the undefined behaviour surfaced by the **Nightly Sanitizer** run
[26861035639](https://github.com/shader-slang/slang/actions/runs/26861035639):

```
source/core/slang-string-util.cpp:550:29: runtime error: member call on address …
which does not point to an object of type 'ISlangBlob'
  note: object is of type 'hlsl::InternalDxcBlobEncoding_Impl<hlsl::DxcBlobUtf8_Impl>'
  #0 StringUtil::getSlice(ISlangBlob*)
  #1 _getSlice(IDxcBlob*)                       source/compiler-core/slang-dxc-compiler.cpp:68
  #2 _handleOperationResult(…)
  #3 DXCDownstreamCompiler::compile(…)
```

## Root cause

DXC's `IDxcBlob` and Slang's `ISlangBlob` are *distinct* COM interfaces that happen
to be ABI/layout-compatible (`IUnknown` + a buffer pointer/size getter at the same
vtable slots). `slang-dxc-compiler.cpp` reinterpret-casts `IDxcBlob*` → `ISlangBlob*`
and calls `ISlangBlob`'s virtuals on the result. It works at runtime, but it is UB —
the object isn't actually an `ISlangBlob` — and UBSan's `-fsanitize=vptr` flags it.

**Why it started failing now:** #10935 ("Build DXC from source when system GLIBC does
not match") means DXC is now compiled inside the sanitizer-instrumented build, so
UBSan has RTTI for the DXC blob types and can finally see the type mismatch that was
invisible while DXC was a prebuilt binary. The CI triage confirms it: *"1 PR-related,
0 pre-existing."*

## Fix

Add a small `DxcBlob` adapter — a genuine `ISlangBlob` (deriving from `BlobBase`) that
holds a ref to the `IDxcBlob` and forwards `getBufferPointer`/`getBufferSize` to its
`GetBufferPointer`/`GetBufferSize`. Use it at the flagged sites: `_getSlice` and the
three `addRepresentationUnknown` calls. Holding the ref also keeps the DXC blob's data
alive for stored artifact representations (the old pun relied on the caller's `ComPtr`).

## Scope / follow-up

The `IDxcLinker::RegisterLibrary` path (`libraryBlobs` … `(IDxcBlob*)libraryBlobs[i].get()`)
round-trips blobs `ISlangBlob* ↔ IDxcBlob*` in both directions. De-punning it requires
creating a real DXC blob via `IDxcLibrary` (not in scope in that function), so it's left
as a follow-up.

## Validation

I could not build/run this locally (macOS, no DXC/DXIL toolchain). I verified the
*mechanism* with a standalone repro: the reinterpret-pun trips `-fsanitize=vptr` with
the identical diagnostic, while the adapter form is clean. **Relying on this PR's
sanitizer CI to validate the actual build** — the branch was pushed to the main repo so
the Nightly Sanitizer can run against it.

### CODE REVIEW

File: source/compiler-core/slang-dxc-compiler.cpp [hunk 1]
```diff
@@ -63,9 +63,46 @@ namespace Slang
...
+// DXC's `IDxcBlob` and Slang's `ISlangBlob` are layout-compatible COM
+// interfaces (both `IUnknown` plus a buffer pointer/size getter), and the code
+// here used to reinterpret-cast an `IDxcBlob*` to `ISlangBlob*` and call
+// `ISlangBlob`'s virtuals on it. That is undefined behaviour: the object is not
+// actually an `ISlangBlob`, and UBSan's `-fsanitize=vptr` flags it (member call
+// on an object that is not of type `ISlangBlob`) once DXC is built from source
+// and is visible to the sanitizer. This adapter keeps a reference to the
+// underlying `IDxcBlob` and forwards the accessors, so the data also outlives
+// any artifact representation that stores the blob.
+class DxcBlob : public BlobBase
```
Comment (jvepsalainen-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:96]: The PR description's "Scope / follow-up" section names the `IDxcLinker::RegisterLibrary` round-trip (lines 691 and 711) as a remaining UB site, and after the patch the reverse-direction pun at `DxcIncludeHandler::LoadSource` (line 155, `*outSource = (IDxcBlob*)blob.detach();`) is also still there with its long-standing `// NOTE! This only works because ISlangBlob is *binary compatible* with IDxcBlob` comment unchanged. After merge, the PR description is detached from the source; a maintainer auditing UBSan failures from this file won't see that some `IDxcBlob ↔ ISlangBlob` puns were intentionally left out of scope. Consider adding a brief in-code `TODO`/`NOTE` block — either in the new `DxcBlob` class doc or alongside the remaining cast sites — naming the three remaining puns and the reason de-punning them is harder. The code stays as-is.

Comment (jvepsalainen-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:96]: The class docblock explains the UB this adapter exists to *eliminate* in the current call sites, but doesn't state two contracts a future maintainer ends up having to infer: (1) the going-forward rule — "anywhere you would have written `(ISlangBlob*)dxcBlob`, write `DxcBlob::create(dxcBlob)` instead; do not construct `DxcBlob` directly (the constructor is private)"; and (2) the encapsulation choice in the factory — it deliberately returns `ComPtr<ISlangBlob>` (not `ComPtr<DxcBlob>`) so callers can't reach the wrapped `IDxcBlob` through the adapter. A short "Usage:" paragraph in the class block covering both would lock the contract in and explain the factory return type at the same time. Without it, a maintainer skimming the class might re-introduce a `reinterpret_cast` somewhere, or be surprised the factory doesn't return its own type.

File: source/compiler-core/slang-dxc-compiler.cpp [hunk 2]
```diff
@@ -63,9 +63,46 @@ namespace Slang
...
+        : m_blob(blob)
+    {
+    }
+    ComPtr<IDxcBlob> m_blob;
+};
+
 static UnownedStringSlice _getSlice(IDxcBlob* blob)
 {
-    return StringUtil::getSlice((ISlangBlob*)blob);
+    return StringUtil::getSlice(DxcBlob::create(blob));
```
Comment (jvepsalainen-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:127]: `DxcBlob::create(blob)` creates a `ComPtr<ISlangBlob>` temporary whose lifetime ends at the function-return semicolon. `StringUtil::getSlice` returns an `UnownedStringSlice` that points into the wrapped buffer. The slice therefore stays valid only because the caller of `_getSlice` retains its own `IDxcBlob` ref (the `DxcBlob` temporary's destruction releases the *adapter's* ref but not the caller's). This is the same lifetime contract as the old reinterpret-pun version — equally implicit, and equally fragile if a caller is ever rewritten to drop its `ComPtr` before consuming the slice. A one-line comment ("returned slice is valid for as long as the caller's `blob` ref outlives it") on `_getSlice` would make the contract visible without re-deriving it from the `StringUtil::getSlice` signature.

File: source/compiler-core/slang-dxc-compiler.cpp [hunk 3]
```diff
@@ -63,9 +63,46 @@ namespace Slang
...
+    }
+    SLANG_NO_THROW size_t SLANG_MCALL getBufferSize() SLANG_OVERRIDE
+    {
+        return m_blob->GetBufferSize();
+    }
+
+    static ComPtr<ISlangBlob> create(IDxcBlob* blob)
+    {
+        return ComPtr<ISlangBlob>(new DxcBlob(blob));
+    }
```
Comment (jvepsalainen-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:114]: `DxcBlob::create` takes an `IDxcBlob*` without asserting non-null; the wrapped accessors would dereference `m_blob` unconditionally if called on a null-backed adapter. Today the three `compile`/`convert` call sites either pre-check (`if (dxcResultBlob)`) or come from a `SLANG_RETURN_ON_FAIL` path that wouldn't produce a null blob, so this is safe; but a future caller adding a fourth site has no in-code reminder that `nullptr` is not allowed. Either `SLANG_ASSERT(blob)` in the constructor or a one-line "blob must be non-null; null wrapping is not supported" in the doc would make the precondition explicit. Low-priority; the existing call sites are correct.

File: source/compiler-core/slang-dxc-compiler.cpp (hunk 1 above)
Comment (jvepsalainen-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:96]: Filling in "A `DxcBlob` is …" cleanly yields "an `ISlangBlob` that adapts an `IDxcBlob`" — the class's actual concept is the adaptation, not the underlying blob. The first line of the doc agrees ("Adapts a DXC `IDxcBlob` to a genuine Slang `ISlangBlob`"), so the name slightly under-sells the role. A name like `DxcBlobAdapter`, `SlangBlobFromDxcBlob`, or similar would match the doc and avoid the slightly misleading "this *is* the DXC blob" read. The class is file-local so the cost of a longer name is low; raising as a possibility, not a must-fix.

**Review by jvepsalainen-nv — COMMENTED** [review]
Claude Opus 4.7 authored clarity review:

A small, well-scoped UB fix (+43/−5 in one file). The PR description does a good job
of explaining the root cause (DXC built in-tree → UBSan can now see the vptr
mismatch on `(ISlangBlob*)dxcBlob`-style puns) and acknowledges its own scope — the
`IDxcLinker::RegisterLibrary` round-trip is left as follow-up. The diff matches the
description and reads cleanly; the new `DxcBlob` adapter is the obvious right shape
for the change.

The clarity feedback below is all about getting a little more durable documentation
into the source while the new declarations are still fresh. None of it is blocking.

> ## Main Concerns

> 1. The "Scope / follow-up" context in the PR description gets detached from the
>    source after merge. The reverse-direction pun at
>    `DxcIncludeHandler::LoadSource` (`*outSource = (IDxcBlob*)blob.detach();`) and
>    the `IDxcLinker::RegisterLibrary` round-trip both stay in the file with their
>    existing comments unchanged. A short in-code TODO/NOTE block — either in the
>    new `DxcBlob` docblock or next to the remaining cast sites — would preserve
>    the "we know about these; they're hard to fix here" reasoning for the next
>    person who hits UBSan on this file.

> 2. The new `DxcBlob` docblock explains the past (the UB the adapter eliminates)
>    but not the going-forward rule. A future maintainer adding a fifth
>    `IDxcBlob → ISlangBlob` site can't tell from the class alone that
>    `DxcBlob::create()` is the only legal path, or why the factory returns
>    `ComPtr<ISlangBlob>` rather than `ComPtr<DxcBlob>`. Both contracts deserve a
>    line in the docblock.

> ## Minor / Nits

> - `_getSlice`'s returned `UnownedStringSlice` is valid only while the caller's
>   `blob` ref outlives it. Same as before the change, but the PR is the natural
>   place to capture the lifetime contract since `_getSlice`'s body just shifted.
> - `DxcBlob::create` takes `IDxcBlob*` without a non-null precondition;
>   `SLANG_ASSERT` or a one-line doc would lock that in while the factory is
>   small.
> - The class name itself reads as "this *is* a DXC blob" rather than "an adapter
>   over one." Low-confidence styling note; the docblock's first line already
>   says "Adapts a DXC `IDxcBlob` …", so the doc and the name are slightly out of
>   sync.

Inline comments below for the four anchors.

File: source/compiler-core/slang-dxc-compiler.cpp (hunk 1 above)
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-dxc-compiler.cpp:96]: Addressed in 5c0df9f. Added a TODO block to the `DxcBlob` docblock that enumerates the three remaining puns (`LoadSource` outparam and the `IDxcLinker::RegisterLibrary` round-trip) and why each is harder to de-pun, so the context survives the merge.

Comment (jvepsalainen-nv) [reply source/compiler-core/slang-dxc-compiler.cpp:96]: Addressed in 5c0df9f. Added a Usage paragraph to the `DxcBlob` docblock: callers must use `DxcBlob::create()` (constructor private), and the factory deliberately returns `ComPtr<ISlangBlob>` rather than `ComPtr<DxcBlob>` to keep the wrapped `IDxcBlob` encapsulated.

File: source/compiler-core/slang-dxc-compiler.cpp (hunk 2 above)
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-dxc-compiler.cpp:127]: Intentionally skipping for this PR. The lifetime contract on `_getSlice` is unchanged from the pre-patch reinterpret-pun version (the caller's `blob` ref already had to outlive the returned slice). Worth a one-liner if anyone is in the area for a follow-up, but it's not a clarity gap this PR created.

File: source/compiler-core/slang-dxc-compiler.cpp (hunk 3 above)
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-dxc-compiler.cpp:114]: Addressed in 5c0df9f. Added `SLANG_ASSERT(blob);` to the private constructor — the factory signature is unchanged, the precondition is now enforced in debug builds, and the assertion documents the contract for future callers.

File: source/compiler-core/slang-dxc-compiler.cpp (hunk 1 above)
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-dxc-compiler.cpp:96]: Intentionally skipping for this PR. The Usage paragraph added to the class docblock in 5c0df9f already establishes the adapter framing in prose; renaming `DxcBlob` -> `DxcBlobAdapter` would touch 6 call sites in a file the PR is otherwise localising tightly. Happy to revisit in a follow-up if the name continues to feel misleading.

File: source/compiler-core/slang-dxc-compiler.cpp [hunk 4]
```diff
@@ -63,9 +63,66 @@ namespace Slang
...
+        return m_blob->GetBufferPointer();
+    }
+    SLANG_NO_THROW size_t SLANG_MCALL getBufferSize() SLANG_OVERRIDE
+    {
+        return m_blob->GetBufferSize();
+    }
+
+    static ComPtr<ISlangBlob> create(IDxcBlob* blob)
+    {
+        return ComPtr<ISlangBlob>(new DxcBlob(blob));
```
Comment (skiminki-nv) [inline source/compiler-core/slang-dxc-compiler.cpp:113]: This should probably do a null check here. I'd probably do both
- assert for non-NULL (private constructor should be ok)
- `if (blob) return ComPtr<ISlangBlob>(new DxcBlob(blob)); else return nullptr;`

**Review by skiminki-nv — DISMISSED** [review]

**Review by skiminki-nv — APPROVED** [review]
